fairscale/optim/adascale.py

```diff
@@ -387,6 +387,44 @@ def _update_avg(self, name: str, value: np.ndarray, factor: float) -> None:
             else:
                 self._state[name] = factor * self._state[name] + (1.0 - factor) * value
 
+    def _gather_flat_grad(self) -> torch.Tensor:
+        """
+        Helper function for gathering all gradients into a single vector.
+        Duplicated from torch.optim.lbfgs.
+        """
+        views = []
+        for param_group in self._optimizer.param_groups:
+            for p in param_group["params"]:
+                if p.grad is None:
+                    view = p.new(p.numel()).zero_()
+                elif p.grad.is_sparse:
+                    view = p.grad.to_dense().view(-1)
+                else:
+                    view = p.grad.view(-1)
+                views.append(view)
+        return torch.cat(views, 0)
+
+    def _compute_intra_grad_corr_mean(self) -> float:
+        """
+        Helper function for computing average intra correlation among gradients on different GPUs.
+        """
+        assert self._world_size > 1, "Only for distributed training"
+        flat_grad = self._gather_flat_grad()
+        corr_mean = torch.tensor(0.0).cuda()
+        if dist.get_rank() == 0:
+            size = flat_grad.numel()
+            gathered_tensors = [torch.zeros(size, device=0) for _ in range(self._world_size)]
+            dist.gather(flat_grad, gather_list=gathered_tensors, dst=0)
+            # the following requires torch 1.10+
+            corr = torch.stack(gathered_tensors).corrcoef()  # type: ignore
+            # pick out the upper triangular part of the correlation matrix
+            corr = corr[torch.triu(torch.ones_like(corr), diagonal=1) == 1]
+            corr_mean = corr.mean()
+        else:
+            dist.gather(flat_grad, gather_list=None, dst=0)
+        dist.broadcast(corr_mean, src=0)
+        return corr_mean.item()
```

Comment:
.item() call would introduce a cpu-gpu sync. Is it actually needed here? I feel that it should be the caller does it in case it needs the float value. Otherwise, just let this function return the tensor to allow async execution of the cpu thread is better.